diff --git a/pappl/job-process.c b/pappl/job-process.c
index ab15f77..b246565 100644
--- a/pappl/job-process.c
+++ b/pappl/job-process.c
@@ -64,8 +64,6 @@ papplJobCreatePrintOptions(
   if ((options = calloc(1, sizeof(pappl_pr_options_t))) == NULL)
     return (NULL);
 
-  options->media = printer->driver_data.media_default;
-
   pthread_rwlock_rdlock(&printer->rwlock);
 
   // copies
@@ -105,7 +103,10 @@ papplJobCreatePrintOptions(
   }
 
   // media-xxx
-  options->media = printer->driver_data.media_default;
+  for (i = 0; i < printer->driver_data.num_source; i ++)
+    if (strcmp(printer->driver_data.source[i], printer->driver_data.media_default.source) == 0)
+      break;
+  options->media = printer->driver_data.media_ready[i];
 
   if ((attr = ippFindAttribute(job->attrs, "media-col", IPP_TAG_BEGIN_COLLECTION)) != NULL)
   {
