diff --git a/pappl/printer-driver.c b/pappl/printer-driver.c
index 26a1ee7..721965a 100644
--- a/pappl/printer-driver.c
+++ b/pappl/printer-driver.c
@@ -297,6 +297,8 @@ papplPrinterSetReadyMedia(
     int               num_ready,	// I - Number of ready media
     pappl_media_col_t *ready)		// I - Array of ready media
 {
+  int i;
+
   if (!printer || num_ready <= 0 || !ready)
     return (false);
 
@@ -310,6 +312,19 @@ papplPrinterSetReadyMedia(
 
   memset(printer->driver_data.media_ready, 0, sizeof(printer->driver_data.media_ready));
   memcpy(printer->driver_data.media_ready, ready, (size_t)num_ready * sizeof(pappl_media_col_t));
+
+  for (i = 0; i < printer->driver_data.num_source; i ++)
+  {
+    if (!strcmp(printer->driver_data.media_default.source,
+		printer->driver_data.source[i]))
+    {
+      printer->driver_data.media_default = printer->driver_data.media_ready[i];
+      break;
+    }
+  }
+  if (i >= printer->driver_data.num_source)
+    printer->driver_data.media_default = printer->driver_data.media_ready[0];
+
   printer->state_time = time(NULL);
 
   pthread_rwlock_unlock(&printer->rwlock);
